### ISSUE TITLE
How do shaders currently work?

### ISSUE THREAD

**Issue by billywade** [minetest/minetest#7730]
I just need someone to ELI5 this for me, because I don't know a lot about either C++ or shaders. 

So I've noticed that other FLOSS projects like Retroarch and Ishiiruka Dolphin have GLSL shaders like Minetest does. Since the specification is standard, I wanted to ask:

1. Why wouldn't these work just dropped in with the current shaders?

2. These both seem to support importing shaders, even though they're compiled like Minetest. If a frontend interface were to be built for shader selection, is there anything in Irrlicht or anything else that would prevent this from working?

This has been a community request for a while, and because I don't know enough about it, right now it looks like something simple enough for me to hammer out over the course of a couple of weeks, in spite of my zero knowledge. Adjust my expectations, please.

**Comment by celeron55** [comment]
A shader is mostly a programmable filter; it has certain inputs and outputs, many of which depend on the program that is feeding it and then using the result.

So they can't be interchanged, because each project can have different inputs (usually) and outputs (usually not so much) for the shaders, either because they're rendering different things based on different data and different pre-processing steps and whatever, or because they just were designed differently otherwise.

I don't know what exactly Retroarch or Ishiiruka Dolphin is using their shaders for, or what exactly they're giving to them as inputs or outputs.

Minetest most certainly is not feeding its shaders anything similar to any other program though. MT does some fairly weird (pseudo) lighting stuff with different vertex color components and then sets some uniforms (basically a global variable in shader terminology) to mix them into the final result, and then does not do any post-process shader step after that.

**Comment by billywade** [comment]
@celeron55 Thanks for responding. All of their shaders are for post processing, since they both have to emulate a lot of disparate software (and in Retroarch's case hardware). From what I can tell, they both feed the textures and the final output image to the shaders to get texture smoothing and fancy lighting. So in order for something like this to work, a new system of using postprocessing shaders would have to be implemented?

**Comment by paramat** [comment]
Any further replies? Will close in a week.

**Comment by paramat** [comment]
Discussion can continue but no point this being open. billywade you could discuss shaders on MT IRC channels or in the forum.

**Comment by billywade** [comment]
That's fine, I'm learning shaders right now and I'm going to hopefully contribute to an Irrlicht fork later to add in shaders.
